bolt-sidecar/src/state/pricing.rs

```diff
@@ -0,0 +1,271 @@
+    }
+
+    /// Calculate the minimum priority fee for a preconfirmation based on
+    /// https://research.lido.fi/t/a-pricing-model-for-inclusion-preconfirmations/9136
+    ///
+    /// # Arguments
+    /// * `incoming_gas` - Gas required by the incoming transaction
+    /// * `preconfirmed_gas` - Total gas already preconfirmed
+    ///
+    /// # Returns
+    /// * `Ok(f64)` - The minimum priority fee in Wei
+    /// * `Err(PricingError)` - If the calculation cannot be performed
+    pub fn calculate_min_priority_fee(
+        &self,
+        incoming_gas: u64,
+        preconfirmed_gas: u64,
+    ) -> Result<u64, PricingError> {
+        // Check if preconfirmed gas exceeds block limit
+        if preconfirmed_gas >= self.block_gas_limit {
+            return Err(PricingError::ExceedsBlockLimit(preconfirmed_gas, self.block_gas_limit));
+        }
+
+        // Validate incoming gas
+        if incoming_gas == 0 {
+            return Err(PricingError::InvalidGasLimit { incoming_gas });
+        }
+
+        // Check if there is enough gas remaining in the block
+        let remaining_gas = self.block_gas_limit - preconfirmed_gas;
+        if incoming_gas > remaining_gas {
+            return Err(PricingError::InsufficientGas {
+                requested: incoming_gas,
+                available: remaining_gas,
+            });
+        }
+
+        // T(IG,UG) = 0.019 * ln(1.02⋅10^-6(30M-UG)+1 / 1.02⋅10^-6(30M-UG-IG)+1) / IG
+        // where
+        // IG = Gas used by the incoming transaction
+        // UG = Gas already preconfirmed
+        // T = Inclusion tip per gas
+        // 30M = Current gas limit (36M soon?)
+        let after_gas = remaining_gas - incoming_gas;
+
+        // Calculate numerator and denominator for the logarithm
+        let numerator = self.gas_scalar * (remaining_gas as f64) + 1.0;
+        let denominator = self.gas_scalar * (after_gas as f64) + 1.0;
+
+        // Calculate gas used
+        let expected_val = self.base_multiplier * (numerator / denominator).ln();
+
+        // Calculate the inclusion tip
+        let inclusion_tip_ether = expected_val / (incoming_gas as f64);
+        let inclusion_tip_wei = (inclusion_tip_ether * 1e18) as u64;
+
+        Ok(inclusion_tip_wei)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_min_priority_fee_zero_preconfirmed() {
+        let pricing = PreconfPricing::default();
+
+        // Test minimum fee (21k gas ETH transfer, 0 preconfirmed)
+        let incoming_gas = 21_000;
+        let preconfirmed_gas = 0;
+        let min_fee_wei =
+            pricing.calculate_min_priority_fee(incoming_gas, preconfirmed_gas).unwrap();
+
+        // Pricing model article expects fee of 0.61 Gwei
+        assert!(
+            (min_fee_wei as f64 - 613_499_092.0).abs() < 1_000.0,
+            "Expected ~613,499,092 Wei, got {} Wei",
+            min_fee_wei
+        );
+    }
```

Comment:
We should find out why we're getting a 0.02 gwei difference between this code and their model. Perhaps working in wei instead of ether can help

Reply:
Won't look into it now, but created an issue to address it later
https://github.com/chainbound/bolt/issues/596